### PR TITLE
Refactor Vision DataModules

### LINKED ISSUE #395: Refactor Vision DataModules
## 🚀 Feature
I was going through the smaller vision data modules (CIFAR10, MNIST) and noticed that there is a lot of duplicate code in them. I would suggest creating a VisionDataModule similar to PyTorch's VisionDataset and create subclasses of it for the different data modules.

### Motivation
Makes maintaining the code easier.

If you agree, I could prepare a PR :)

### PR DESCRIPTION
**Pull request by chris-clem**
## What does this PR do?

As proposed in #395, this PR creates a `BaseDataModule`, that can be used as the parent class for several existing DataModules like `BinaryMNISTDataModule`, `CIFAR10DataModule`, `FashionMNISTDataModule`, and `MNISTDataModule`.

This removes duplicate code in the DataModules and thus makes maintaining them easier.

Fixes #395 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_bolts/datamodules/base_datamodule.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    def __init__(
+        self,
+        dataset_cls,
+        dims: Tuple[int, int, int],
+        data_dir: Optional[str] = None,
+        val_split: int = 5000,
+        num_workers: int = 16,
+        normalize: bool = False,
+        seed: int = 42,
+        batch_size: int = 32,
```
Comment (nateraw) [inline pl_bolts/datamodules/base_datamodule.py:32]: I like the idea of `eval_batch_size` and `train_batch_size` being separate

File: pl_bolts/datamodules/base_datamodule.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+class BaseDataModule(LightningDataModule):
+
+    extra_args = {}
+
+    def __init__(
+        self,
+        dataset_cls,
+        dims: Tuple[int, int, int],
+        data_dir: Optional[str] = None,
+        val_split: int = 5000,
```
Comment (nateraw) [inline pl_bolts/datamodules/base_datamodule.py:28]: Should this be an integer, a float, or perhaps optionally both? I think intuitively a float makes most sense, as you wont have to hard-code for future datasets.

but this is fine as-is too.

File: pl_bolts/datamodules/base_datamodule.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        self.dataset_cls(
+            self.data_dir, train=False, download=True, transform=transform_lib.ToTensor(), **self.extra_args
+        )
+
+    def train_dataloader(self):
+        """
+        Train set removes a subset to use for validation
+        """
+        transforms = self.default_transforms() if self.train_transforms is None else self.train_transforms
+        dataset = self.dataset_cls(self.data_dir, train=True, download=False, transform=transforms, **self.extra_args)
```
Comment (nateraw) [inline pl_bolts/datamodules/base_datamodule.py:71]: why doesnt the logic of initializing and splitting the dataset live in the `setup` hook?

File: pl_bolts/datamodules/base_datamodule.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            pin_memory=True,
+        )
+        return loader
+
+    def val_dataloader(self):
+        """
+        Val set uses a subset of the training set for validation
+        """
+        transforms = self.default_transforms() if self.val_transforms is None else self.val_transforms
+        dataset = self.dataset_cls(self.data_dir, train=True, download=False, transform=transforms, **self.extra_args)
```
Comment (nateraw) [inline pl_bolts/datamodules/base_datamodule.py:91]: same as above comment. this should live in `setup`.

**Review by nateraw — COMMENTED** [review]
nice work 😄 . couple notes.

File: pl_bolts/datamodules/base_datamodule.py (hunk 1 above)
Comment (chris-clem) [reply pl_bolts/datamodules/base_datamodule.py:32]: This would break a lot of tests. Should I still do it?

**Comment by Borda** [comment]
@ananyahjha93 @nateraw is it ready to go?

**Comment by chris-clem** [comment]
Any updates?

File: pl_bolts/datamodules/base_datamodule.py (hunk 1 above)
Comment (Borda) [reply pl_bolts/datamodules/base_datamodule.py:33]: Let's move this change to separate PR?
in theory, how much the train and valid batch size can have an effect?

**Review by akihironitta — APPROVED** [review]
@chris-clem It looks good to me.
@nateraw @ananyahjha93 Can you also have a look?

**Comment by Borda** [comment]
@chris-clem can you pls resolve conflits?

**Comment by chris-clem** [comment]
> @chris-clem can you pls resolve conflits?

I just did it

**Comment by Borda** [comment]
@chris-clem we have merged better importing with if/else instead of try/except mind pls update the PR..
sorry for the inconvenience, but I think it shall land soon :]

File: pl_bolts/datamodules/base_datamodule.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+try:
+    from torchvision import transforms as transform_lib
+except ModuleNotFoundError:
+    warn_missing_pkg("torchvision")  # pragma: no-cover
+    _TORCHVISION_AVAILABLE = False
+else:
+    _TORCHVISION_AVAILABLE = True
+
+
+class BaseDataModule(LightningDataModule):
```
Comment (Borda) [inline pl_bolts/datamodules/base_datamodule.py:20]: if it is refactor for Vision lets call it so and do not confuse users with another base level
```suggestion
class VisionDataModule(LightningDataModule):
```

Comment (Borda) [reply pl_bolts/datamodules/base_datamodule.py:20]: btw, is it any vision or `torchvision` only?

**Review by Borda — COMMENTED** [review]
could we pls also add typing?

File: pl_bolts/datamodules/base_datamodule.py (hunk 5 above)
Comment (chris-clem) [reply pl_bolts/datamodules/base_datamodule.py]: It is for the basic torchvision datasets with the same API (CIFAR10, FashionMNIST, MNIST)

**Comment by chris-clem** [comment]
> @chris-clem we have merged better importing with if/else instead of try/except mind pls update the PR..
> sorry for the inconvenience, but I think it shall land soon :]

No problem, I updated the PR

File: pl_bolts/datamodules/base_datamodule.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        """
+        Creates train, val, and test dataset
+        """
+        if stage == "fit" or stage is None:
+            train_transforms = (
+                self.default_transforms() if self.train_transforms is None else self.train_transforms
+            )
+            val_transforms = (
+                self.default_transforms() if self.val_transforms is None else self.val_transforms
+            )
```
Comment (Borda) [inline pl_bolts/datamodules/base_datamodule.py:67]:
```suggestion
            train_transforms = self.default_transforms() if self.train_transforms is None else self.train_transforms
            val_transforms = self.default_transforms() if self.val_transforms is None else self.val_transforms
```

File: pl_bolts/datamodules/base_datamodule.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+            dataset_train = self.dataset_cls(self.data_dir, train=True, transform=train_transforms, **self.extra_args)
+            dataset_val = self.dataset_cls(self.data_dir, train=True, transform=val_transforms, **self.extra_args)
+
+            # Split
+            self.dataset_train = self._split_dataset(dataset_train)
+            self.dataset_val = self._split_dataset(dataset_val, train=False)
+
+        if stage == "test" or stage is None:
```
Comment (Borda) [inline pl_bolts/datamodules/base_datamodule.py:76]:
```suggestion
        elif stage == "test" or stage is None:
```

File: pl_bolts/datamodules/binary_mnist_datamodule.py [hunk 8]
```diff
@@ -44,123 +42,61 @@ class BinaryMNISTDataModule(LightningDataModule):
...
+            dims=dims,
+            data_dir=data_dir,
+            val_split=val_split,
+            num_workers=num_workers,
+            normalize=normalize,
+            seed=seed,
+            batch_size=batch_size,
+            *args,
+            **kwargs,
+        )
```
Comment (Borda) [inline pl_bolts/datamodules/binary_mnist_datamodule.py:84]:
```suggestion
        super().__init__(
            dataset_cls=BinaryMNIST,
            dims=(1, 28, 28),
            data_dir=data_dir,
            val_split=val_split,
            num_workers=num_workers,
            normalize=normalize,
            seed=seed,
            batch_size=batch_size,
            *args,
            **kwargs,
        )
```

File: pl_bolts/datamodules/cifar10_datamodule.py [hunk 9]
```diff
@@ -58,125 +54,76 @@ class CIFAR10DataModule(LightningDataModule):
         dm.val_transforms  = ...
     """
 
-    name = 'cifar10'
+    name = "cifar10"
```
Comment (Borda) [inline pl_bolts/datamodules/cifar10_datamodule.py:57]:
```suggestion
    NAME = "cifar10"
```

**Review by Borda — COMMENTED** [review]
I will implement some other comments as direct commits here

**Comment by Borda** [comment]
@chris-clem @akihironitta I have made some more suggestions and rebased on master, mind check it...

File: pl_bolts/datamodules/vision_datamodule.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        self,
+        data_dir: Optional[str] = None,
+        val_split: Union[int, float] = 0.2,
+        num_workers: int = 16,
+        normalize: bool = False,
+        seed: int = 42,
+        batch_size: int = 32,
+        *args,
+        **kwargs,
+    ):
```
Comment (akihironitta) [inline pl_bolts/datamodules/vision_datamodule.py:29]: Recently in #388 by @briankosw, we added `pin_memory`, `shuffle` and `drop_last` to all datamodules, so could you include them here, too?

File: pl_bolts/datamodules/cifar10_datamodule.py [hunk 11]
```diff
@@ -58,125 +54,75 @@ class CIFAR10DataModule(LightningDataModule):
...
+            batch_size=batch_size,
+            *args,
+            **kwargs,
+        )
 
     @property
-    def num_classes(self):
+    def num_samples(self) -> int:
+        len_dataset = 50_000
+        # todo: clean up with parent methods
```
Comment (akihironitta) [inline pl_bolts/datamodules/cifar10_datamodule.py:101]: What do you mean by `# todo: clean up with parent methods`?

File: pl_bolts/datamodules/vision_datamodule.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+            val_transforms = self.default_transforms() if self.val_transforms is None else self.val_transforms
+
+            dataset_train = self.DATASET_CLASS(self.data_dir, train=True, transform=train_transforms, **self.EXTRA_ARGS)
+            dataset_val = self.DATASET_CLASS(self.data_dir, train=True, transform=val_transforms, **self.EXTRA_ARGS)
+
+            # Split
+            self.dataset_train = self._split_dataset(dataset_train)
+            self.dataset_val = self._split_dataset(dataset_val, train=False)
+
+        elif stage == "test" or stage is None:
```
Comment (akihironitta) [inline pl_bolts/datamodules/vision_datamodule.py:71]: I think `stage is None` in `elif` is unnecessary in this case:
```python
if stage== "fit" or stage is None:
    ...
elif stage == "test":
    ...
```

or if we want to setup both "fit" and "test" when `stage is None`,  we should write the conditions with `if` without `elif`:
```python
if stage== "fit" or stage is None:
    ...
if stage == "test" or stage is None:
    ...
```
cc: @Borda

**Review by akihironitta — CHANGES_REQUESTED** [review]
@chris-clem @Borda The changes basically look great to me! I left some comments, so could you have a look?

File: pl_bolts/datamodules/cifar10_datamodule.py (hunk 11 above)
Comment (Borda) [reply pl_bolts/datamodules/cifar10_datamodule.py:101]: this seems to be very similar to the one in `VisionDataModule` but has not checked details

File: pl_bolts/datamodules/vision_datamodule.py (hunk 12 above)
Comment (Borda) [reply pl_bolts/datamodules/vision_datamodule.py:71]: good catch

**Comment by Borda** [comment]
@chris-clem mind resolve conflicts and https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/400#discussion_r545036430

File: CHANGELOG.md [hunk 13]
```diff
@@ -14,6 +14,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Added data monitor callbacks `ModuleDataMonitor` and `TrainingDataMonitor` ([#285](https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/285))
 
+- Added `BaseDataModule` as parent class for `BinaryMNISTDataModule`, `CIFAR10DataModule`, `FashionMNISTDataModule`,
```
Comment (chris-clem) [inline CHANGELOG.md:17]:
```suggestion
- Added `VisionDataModule` as parent class for `BinaryMNISTDataModule`, `CIFAR10DataModule`, `FashionMNISTDataModule`, 
```

File: pl_bolts/datamodules/binary_mnist_datamodule.py [hunk 14]
```diff
@@ -41,125 +40,70 @@ class BinaryMNISTDataModule(LightningDataModule):
     """
 
     name = "binary_mnist"
+    dataset_cls = BinaryMNIST
+    dims = (1, 28, 28)
```
Comment (chris-clem) [inline pl_bolts/datamodules/binary_mnist_datamodule.py:44]: @Borda, I changed it to lower case, so `LightningDataModule.size()` works.

File: pl_bolts/datamodules/cifar10_datamodule.py [hunk 15]
```diff
@@ -58,136 +54,75 @@ class CIFAR10DataModule(LightningDataModule):
         dm.val_transforms  = ...
     """
 
-    name = 'cifar10'
-    extra_args = {}
+    if not _TORCHVISION_AVAILABLE:
+        raise ModuleNotFoundError(  # pragma: no-cover
+            "You want to use CIFAR10 dataset loaded from `torchvision` which is not installed yet."
+        )
```
Comment (chris-clem) [inline pl_bolts/datamodules/cifar10_datamodule.py:60]: @Borda, @akihironitta, how can I fix the CI base testing tests without torchvision? I thought moving up the error raising would help but it doesn't.

Comment (Borda) [reply pl_bolts/datamodules/cifar10_datamodule.py:60]: you need to 
```
if _TORCHVISION_AVAILABLE:
    from torchvision... import MNIST
else:
    MNIST = None
```

Comment (chris-clem) [reply pl_bolts/datamodules/cifar10_datamodule.py]: Thanks!

File: pl_bolts/datamodules/vision_datamodule.py [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+        self.val_split = val_split
+        self.num_workers = num_workers
+        self.normalize = normalize
+        self.batch_size = batch_size
+        self.seed = seed
+        self.shuffle = shuffle
+        self.pin_memory = pin_memory
+        self.drop_last = drop_last
+
+    def prepare_data(self):
```
Comment (akihironitta) [inline pl_bolts/datamodules/vision_datamodule.py:59]:
```suggestion
    def prepare_data(self) -> None:
```

File: pl_bolts/datamodules/vision_datamodule.py [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+        """
+        Saves files to data_dir
+        """
+        self.dataset_cls(self.data_dir, train=True, download=True)
+        self.dataset_cls(self.data_dir, train=False, download=True)
+
+    def setup(self, stage: Optional[str] = None):
```
Comment (akihironitta) [inline pl_bolts/datamodules/vision_datamodule.py:66]:
```suggestion
    def setup(self, stage: Optional[str] = None) -> None:
```

**Review by akihironitta — APPROVED** [review]
@chris-clem It looks valid to me :]

**Review by Borda — APPROVED** [review]
